### PR TITLE
Introduce ResourceTypeRegistry that can be configured by the application

### PR DESCRIPTION
**Pull request by Stummi**

### CODE REVIEW

File: spring-boot-starter-scim2/src/main/java/com/bettercloud/scim2/server/resourcetypes/ResourceTypeRegistry.java [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+import com.bettercloud.scim2.server.ResourceTypeDefinition;
+
+import java.util.Set;
+
+public interface ResourceTypeRegistry {
+
+    Set<ResourceTypeDefinition> getResourceTypeDefinitions();
+
+}
```
Comment (Stummi) [inline spring-boot-starter-scim2/src/main/java/com/bettercloud/scim2/server/resourcetypes/ResourceTypeRegistry.java:11]:
```suggestion
}
```

File: spring-boot-starter-scim2/src/main/java/com/bettercloud/scim2/server/controller/discovery/ResourceTypesController.java [hunk 2]
```diff
@@ -77,7 +78,7 @@
 public class ResourceTypesController extends SchemaAwareController {
 
     @Autowired
-    public ResourceTypesController(final Scim2Properties scim2Properties, final Set<ResourceTypeDefinition> resourceDefinitions) {
+    public ResourceTypesController(final Scim2Properties scim2Properties, final ResourceTypeRegistry resourceDefinitions) {
```
Comment (matthiastz) [inline spring-boot-starter-scim2/src/main/java/com/bettercloud/scim2/server/controller/discovery/ResourceTypesController.java]: nit: when changing the type of the var please also rename it

File: spring-boot-starter-scim2/src/main/java/com/bettercloud/scim2/server/controller/discovery/SchemaAwareController.java [hunk 3]
```diff
@@ -23,21 +24,17 @@
...
 
     protected abstract List<GenericScimResource> getResources(final Set<ResourceTypeDefinition> resourceDefinitions);
 
-    protected Set<ResourceTypeDefinition> resourceDefinitions;
+    protected ResourceTypeRegistry resourceDefinitions;
 
     @Autowired
     public SchemaAwareController(final Scim2Properties scim2Properties,
-                                 final Set<ResourceTypeDefinition> resourceDefinitions) {
+                                 final ResourceTypeRegistry resourceDefinitions) {
```
Comment (matthiastz) [inline spring-boot-starter-scim2/src/main/java/com/bettercloud/scim2/server/controller/discovery/SchemaAwareController.java]: same here pls

File: spring-boot-starter-scim2/src/main/java/com/bettercloud/scim2/server/controller/discovery/SchemaAwareController.java [hunk 4]
```diff
@@ -23,21 +24,17 @@
...
 public abstract class SchemaAwareController extends BaseResourceController<GenericScimResource> {
-
-    private List<GenericScimResource> resources;
-
     private final SchemaAwareFilterEvaluator filterEvaluator = new SchemaAwareFilterEvaluator(resourceTypeDefinition);
 
     protected abstract List<GenericScimResource> getResources(final Set<ResourceTypeDefinition> resourceDefinitions);
 
-    protected Set<ResourceTypeDefinition> resourceDefinitions;
+    protected ResourceTypeRegistry resourceDefinitions;
```
Comment (matthiastz) [inline spring-boot-starter-scim2/src/main/java/com/bettercloud/scim2/server/controller/discovery/SchemaAwareController.java]: and here

File: spring-boot-starter-scim2/src/main/java/com/bettercloud/scim2/server/controller/discovery/SchemasController.java [hunk 5]
```diff
@@ -56,7 +57,7 @@
 public class SchemasController extends SchemaAwareController {
 
     @Autowired
-    public SchemasController(final Scim2Properties scim2Properties, final Set<ResourceTypeDefinition> resourceDefinitions) {
+    public SchemasController(final Scim2Properties scim2Properties, final ResourceTypeRegistry resourceDefinitions) {
```
Comment (matthiastz) [inline spring-boot-starter-scim2/src/main/java/com/bettercloud/scim2/server/controller/discovery/SchemasController.java]: and here^^

File: spring-boot-starter-scim2/src/main/java/com/bettercloud/scim2/server/controller/discovery/ResourceTypesController.java (hunk 2 above)
Comment (matthiastz) [inline spring-boot-starter-scim2/src/main/java/com/bettercloud/scim2/server/controller/discovery/ResourceTypesController.java]:
```suggestion
    public ResourceTypesController(final Scim2Properties scim2Properties, final ResourceTypeRegistry resourceTypeRegistry) {
```

File: spring-boot-starter-scim2/src/main/java/com/bettercloud/scim2/server/controller/discovery/SchemaAwareController.java (hunk 4 above)
Comment (matthiastz) [inline spring-boot-starter-scim2/src/main/java/com/bettercloud/scim2/server/controller/discovery/SchemaAwareController.java]:
```suggestion
    protected ResourceTypeRegistry resourceTypeRegistry;
```

File: spring-boot-starter-scim2/src/main/java/com/bettercloud/scim2/server/controller/discovery/SchemaAwareController.java (hunk 3 above)
Comment (matthiastz) [inline spring-boot-starter-scim2/src/main/java/com/bettercloud/scim2/server/controller/discovery/SchemaAwareController.java]:
```suggestion
                                 final ResourceTypeRegistry resourceTypeRegistry) {
```

File: spring-boot-starter-scim2/src/main/java/com/bettercloud/scim2/server/controller/discovery/SchemaAwareController.java [hunk 6]
```diff
@@ -55,7 +52,7 @@ public SchemaAwareController(final Scim2Properties scim2Properties,
             @RequestParam(value = ApiConstants.QUERY_PARAMETER_FILTER, required = false) final String filterString) throws ScimException {
 
         final List<GenericScimResource> filteredResources = StringUtils.isEmpty(filterString)
-                                                            ? resources
+                                                            ? getResources(resourceDefinitions.getResourceTypeDefinitions())
```
Comment (matthiastz) [inline spring-boot-starter-scim2/src/main/java/com/bettercloud/scim2/server/controller/discovery/SchemaAwareController.java]:
```suggestion
                                                            ? getResources(resourceTypeRegistry.getResourceTypeDefinitions())
```

File: spring-boot-starter-scim2/src/main/java/com/bettercloud/scim2/server/controller/discovery/SchemaAwareController.java [hunk 7]
```diff
@@ -78,7 +75,7 @@ public ScimResource get(@PathVariable("id") final String id) throws ScimExceptio
     }
 
     private List<GenericScimResource> filterResources(final Filter filter) {
-        return resources.stream().filter(genericScimResource -> {
+        return getResources(resourceDefinitions.getResourceTypeDefinitions()).stream().filter(genericScimResource -> {
```
Comment (matthiastz) [inline spring-boot-starter-scim2/src/main/java/com/bettercloud/scim2/server/controller/discovery/SchemaAwareController.java]:
```suggestion
        return getResources(resourceTypeRegistry.getResourceTypeDefinitions()).stream().filter(genericScimResource -> {
```

File: spring-boot-starter-scim2/src/main/java/com/bettercloud/scim2/server/controller/discovery/SchemasController.java (hunk 5 above)
Comment (matthiastz) [inline spring-boot-starter-scim2/src/main/java/com/bettercloud/scim2/server/controller/discovery/SchemasController.java]:
```suggestion
    public SchemasController(final Scim2Properties scim2Properties, final ResourceTypeRegistry resourceTypeRegistry) {
```

**Review by matthiastz — APPROVED** [review]

**Pull request merged**